### PR TITLE
Data deanonymization

### PR DESCRIPTION
**Pull request by maks-operlejn-ds**
### Description

The feature for pseudonymizing data with ability to retrieve original text (deanonymization) has been implemented. In order to protect private data, such as when querying external APIs (OpenAI), it is worth pseudonymizing sensitive data to maintain full privacy. But then, after the model response, it would be good to have the data in the original form.

I implemented the `PresidioReversibleAnonymizer`, which consists of two parts:

1. anonymization - it works the same way as `PresidioAnonymizer`, plus the object itself stores a mapping of made-up values to original ones, for example:
```
    {
        "PERSON": {
            "<anonymized>": "<original>",
            "John Doe": "Slim Shady"
        },
        "PHONE_NUMBER": {
            "111-111-1111": "555-555-5555"
        }
        ...
    }
```

2. deanonymization - using the mapping described above, it matches fake data with original data and then substitutes it.

Between anonymization and deanonymization user can perform different operations, for example, passing the output to LLM.

### Future works

- **instance anonymization** - at this point, each occurrence of PII is treated as a separate entity and separately anonymized. Therefore, two occurrences of the name John Doe in the text will be changed to two different names. It is therefore worth introducing support for full instance detection, so that repeated occurrences are treated as a single object.
- **better matching and substitution of fake values for real ones** - currently the strategy is based on matching full strings and then substituting them. Due to the indeterminism of language models, it may happen that the value in the answer is slightly changed (e.g. *John Doe* -> *John* or *Main St, New York* -> *New York*) and such a substitution is then no longer possible. Therefore, it is worth adjusting the matching for your needs.
- **Q&A with anonymization** - when I'm done writing all the functionality, I thought it would be a cool resource in documentation to write a notebook about retrieval from documents using anonymization. An iterative process, adding new recognizers to fit the data, lessons learned and what to look out for

### Twitter handle
@deepsense_ai / @MaksOpp

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
per slack: 

only nit: in this example it would be good to set temperature=0 (its weird that the first, not de-anonymized response is different than the second) and maybe make it so the llm response isnt EXACTLY the same as the input (otherwise its unclear if anything is happeneing

**Comment by maks-operlejn-ds** [comment]
> per slack:
> 
> only nit: in this example it would be good to set temperature=0 (its weird that the first, not de-anonymized response is different than the second) and maybe make it so the llm response isnt EXACTLY the same as the input (otherwise its unclear if anything is happeneing

@hwchase17 Done! I came up with another example, let me know what you think :smile:

**Review by hwchase17 — APPROVED** [review]
looks good to me!

File: libs/experimental/langchain_experimental/data_anonymizer/faker_presidio_mapping.py [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
-def get_pseudoanonymizer_mapping() -> Dict[str, Callable]:
+def get_pseudoanonymizer_mapping(seed: Optional[int] = None) -> Dict[str, Callable]:
     try:
         from faker import Faker
     except ImportError as e:
         raise ImportError(
             "Could not import faker, please install it with `pip install Faker`."
         ) from e
 
+    Faker.seed(seed)
```
Comment (nfcampos) [inline libs/experimental/langchain_experimental/data_anonymizer/faker_presidio_mapping.py:13]: This modifies Faker globally, can we set the seed only on the instance we create here

Comment (maks-operlejn-ds) [reply libs/experimental/langchain_experimental/data_anonymizer/faker_presidio_mapping.py]: Done! :smile:

**Comment by maks-operlejn-ds** [comment]
@hwchase17 checks have passed, ready to merge :+1:

**Review by hinthornw — APPROVED** [review]

**Pull request merged**